UAF-149 UAF-1994 Added the route node for Organization Fund Reviewer to the ACCT and GACC Documents

leahasullivan opened the pull request:
spring-coa.xml
- Added the bean definition for the organizationFundReviewRoleTypeService and pointed it to the new OrganizationFundReviewRoleTypeServiceImpl.java that was created.

coa-resources.properties
- Added the error message that for the org review role, either Fund Group Code or Sub Fund Group Code must be specified.

AccountMaintenanceDocument.xml (DataDictionary)
- Added the routingTypeDefinition for the OrganizationFundReview to the workflow attributes.
- Added the bean definition for the Routing Attribute Fund with the qualificationAttributeName of fundGroupCode
- Added the bean definition for the DocumentValuePathGroup ChartOrganizationFund
- Added the bean definition for the routingtype OrganizationFundReview specifying the routingAttributes (including the new one added for Fund) and the new documentValuePathGroup that was created for ChartOrganizationFund

AccountMaintenanceDocument.xml (workflow)
- Added the workflow document to be ingested to the project.

AccountGlobalMaintenanceDocument.xml (DataDictionary)
- Added the workflow attribute routingTypeDefinition for the OrganizationFundReview
- Added the bean definition for the DocumentValuePathGroup ChartOrganizationFund
- Added the DocumentCollectionPath chartOrganizationFund
- Added the bean definition for the routingtype OrganizationFundReview specifying the routingAttributes (including the new one added for Fund) and the new documentValuePathGroup that was created for ChartOrganizationFund

AccountGlobalMaintenanceDocument.xml (workflow)
- Added the workflow document to be ingested to the project.

OrgReviewRolesValueFinder.java
- Created class to override the getKeyValues method to add in the org fund only code and text to the drop down list.

OrgReviewRole.xml
- Added override for the OrgReviewRole-reviewRolesIndicator to point to the OrgReviewRolesValuesFinder that was created.

KFSConstants.java
- Added the constant for the node OrganizationFundReview. Note: From KFS3 to KFS6 there were some changes, The route node names are now in a class called "RouteLevelNames" I added the constant there with the others.  I updated the name of the constant to match the other route node names.
- Added the constants for the drop down list: Code and description text

OrgReviewRole.java
- Added the method isOrgFundReviewRoleIndicator to get the orgFundReviewRoleIndicator attribute.
- Added the condition to return the role names for the Org Fund Reviewer in the isBothReviewRolesIndicator method
- Added the condition to the setRoleNamesAndReviewIndicator method if the isOrgFundReviewRoleIndicator is true to set the indicator on the document for the OrgFund Code

OrgReviewRoleServiceImpl.java
** In KFS3 this code was in the OrgReviewRoleLookupableHelperServiceImpl.java
- Added the method hasOrganizationFundReview to check to see if the document has the route node of OrganizationFundReview.  I had to add this method to the org.kuali version of the file because it would not compile from the edu.arizona package.
- Added the call to the new method hasOrganizationFundReview to the getRolesToConsider that checks to see if the document type selected has the route node.
- (EDU ARIZONA) Added code to getRolesToConsider method to add the OrganizationFundReview to the list of rolesToConsider.
- (EDU ARIZONA) Added code to the getRolesToSaveFor method as another statement that if the org fund code equals the review roles indicator to add the Org Fund Review role name to the list of roles to save for. **This method was previously in the OrgReviewRoleMaintainableImpl.java in KFS3
**Part of the modification was an addition to a method called "parentAndChildrenHaveZeroOrgAndAccountReviewRoles", but this method seems to be non-existant in KFS6

OrgReviewRoleRule.java
- Added if statement to check if the role name matched Organization Fund Reviewer and check if the fund group code and sub fund group code were blank. If so, it would display an error that one of the fields were required for the org review role.

OrgReviewRoleMaintainableImpl.java
- Added override for the getSections method.  Added line of code to get back if the document has OrganziationFundReview indicator. Added two more conditions (hasOrganizationFundReview and hasAccountingOrganizationHierarchy) to the determine if Review type fields should be read only.
**The method "shouldReviewTypesFieldBeReadOnly" doesn't exist in KFS6, but did in KFS3.  Looks like the method was rolled up into the getSections method in KFS6.
- Removed the prepareFieldsCommon method, this was causing the Fund Group Code, Sub-Fund Group Code and Object Sub-Type Code read only after returning a value from the Document Type search. (Not sure why I added this to this file for UAF-542 as I don't see it in the commits)

OrganizationFundReviewRoleTypeServiceImpl.java
- Created OrganizationFundReviewRoleTypeServiceImpl file
- Added the performMatch method which traverses the org hierarchy not the document type hierarchy
- Added the doesFundGroupMatch method to compare the role fund group code to the document fund group code
- Added the doesFundGroupMatch method to compare the role sub-fund group code to the document sub-fund group code
- Added the DocumentTypeService to be used to pull the document type name for the perform match method
- Added the getAttributeDefinition method

OrgReviewRoleService.java
- Added method signature to the interface for the new method added called hasOrganizationFundReview (was added to the OrgReviewRoleServiceImpl.java class in the org.kuali package)

spring-coa-bus-exports.xml
- added the bean for the roleTypeServiceCallback for the OrganizationFundReviewRoleTypeService.  This replaces the KSBServiceExporter that was used in the spring-coa.xml in KFS3 when defining the bean for the role type service.

Perrinek commented on kfs-core/src/main/resources/edu/arizona/kfs/coa/businessobject/datadictionary/OrgReviewRole.xml:
```diff
@@ -41,7 +41,20 @@
-		
+
+  <bean id="OrgReviewRole-reviewRolesIndicator" parent="OrgReviewRole-reviewRolesIndicator-parentBean" />
+  <bean id="OrgReviewRole-reviewRolesIndicator-parentBean" abstract="true" parent="AttributeDefinition">
```
I am wondering why you needed to completely override this attribute definition instead of just the property that you needed to change (control). Why not:
bean id="OrgReviewRole-reviewRolesIndicator" parent="OrgReviewRole-reviewRolesIndicator-parentBean"
	property name="control"
		bean parent="RadioControlDefinition"
			p:valuesFinderClass="org.kuali.kfs.coa.businessobject.options.OrgReviewRolesValuesFinder"
			p:includeKeyInLabel="true" /
	/property
/bean

leahasullivan replied:
I think this was just a case where I copied the entire attribute definition and made the change to the control. I can make the change to only override the control and retest.

Perrinek commented on kfs-core/src/main/java/edu/arizona/kfs/coa/document/validation/impl/OrgReviewRoleRule.java:
```diff
@@ -29,6 +30,51 @@ protected boolean validateFundGroupSubFundGroupCode(OrgReviewRole orr) {
+        }
+        if(StringUtils.isNotEmpty(orr.getRoleMemberRoleName())){
+            if ( StringUtils.equals( KFSConstants.SysKimApiConstants.ACCOUNTING_REVIEWER_ROLE_NAME, orr.getRoleMemberRoleName())
+                    || StringUtils.equals( KFSConstants.SysKimApiConstants.ORGANIZATION_REVIEWER_ROLE_NAME, orr.getRoleMemberRoleName() ) ) {
```
Should Organization Fund Review be added to this if?
|| StringUtils.equals( KFSConstants.SysKimApiConstants.ORGANIZATION_FUND_REVIEWER_ROLE_NAME, orr.getRoleMemberRoleName() )

leahasullivan replied:
This method was reworked and has changed a lot from KFS3 to KFS6.  I would assume that you are correct that this should be added since the others are there.  I will add it and retest.

Perrinek commented on kfs-core/src/main/java/edu/arizona/kfs/coa/document/validation/impl/OrgReviewRoleRule.java:
```diff
@@ -29,6 +30,51 @@ protected boolean validateFundGroupSubFundGroupCode(OrgReviewRole orr) {
+        		putFieldError(KfsKimAttributes.SUB_FUND_GROUP_CODE, "error.member.fundgroup.subfundgroup.onerequired");
+        		valid = false;
+        	}
+        }
```
I am just wondering why this validation (for statement) is not in its own method like the other validation we added (method validateFundGroupSubFundGroupCode). There doesn't seem to be any reason for this validation to be included in method validateRoleMember.

leahasullivan replied:
I think I put it there since that is where it was in KFS3, but you are correct. It could be added to the method validateFundGroupSubFundGroupCode.  I will move it and retest.

Perrinek commented on kfs-core/src/main/java/edu/arizona/kfs/coa/identity/OrganizationFundReviewRoleTypeServiceImpl.java:
```diff
@@ -0,0 +1,119 @@
+        return this.documentTypeService;
+    }
+    
+    public void setDocumentTypeService(DocumentTypeService documentTypeService) {
```
This setter is not needed.

leahasullivan replied:
This is already removed as I don't see it in my local branch.  I will verify once I commit the other changes that this is removed.

ghost commented:
Initial Review complete. Please refer to my comments and previous discussions.

leahasullivan commented:
Changed the placement of the validation in the OrgReviewRoleRule.java to be included in the validateFundGroupSubFundGroupCode method. Updated the validateRoleMember method to include the OrganizationFundReviewer role name. (commit: https://github.com/ua-eas/kfs/pull/110/commits/45a4b2a91f86041a184872883a00eaed06245159)

Missed it in the last commit, but removed the setter in OrganizationFundReviewRoleTypeServiceImpl.java. (https://github.com/ua-eas/kfs/pull/110/commits/d816f9e93ce599375a99a2b97b622c79beb8a49d)

Previous commit updated the OrgReviewRole.xml attribute definition.  https://github.com/ua-eas/kfs/pull/110/commits/dfacc4bf5194779baa3229675d6eeb1dc4230c19

ghost commented:
Looks good!
